packages/address/src/AddressFormatter.ts

```diff
@@ -89,11 +97,13 @@ export default class AddressFormatter {
     });
   }
 
-  private loadCountryFromCache(
-    countryCode: string,
-  ): Country | undefined | null {
-    if (ORDERED_COUNTRIES_CACHE[this.locale]) {
-      return ORDERED_COUNTRIES_CACHE[this.locale].find((country) => {
+  /**
+   * Synchronous API to consult the cached countries before having to resort to
+   * fetching.
+   */
+  loadCountryFromCache(countryCode: string): Country | undefined | null {
```

Comment:
I don't understand why we'd need to make this public. What is the the use case for needing to access the value from the cache (and returning null if its not found in the cache is empty) instead of calling `getCountry()`

Reply:
Returning `null` or `undefined` was left as-is, not to change the internal API. I guess `null` could mean the cache isn't set and `undefined` could mean the `countryCode` wasn't found. This could be either documented or changed so that it's clearer.

As for the motivation behind the change: since `getCountry` is async, it forces an async behavior all the time even though a sync usage could be possible after the first load. This forces a flash of skeleton in React components even when all the data is already cached.

Reply:
> Returning null or undefined was left as-is, not to change the internal API. 

I get that, but I'm not sure why its useful to expose this function. Exposing this publicly we have two mechanisms of getting a country - getCountry and  loadCountryFromCache - one of which shall always return correct data, and the other is a footgun that won't work unless a cache is already primed. That's difficult to understand and teach and leaves room for errors - people will become confused why sometimes their content load and sometimes it doesn't if they use loadCountryFromCache.


Can you please expand on why you believe you need this. I wonder if there's is a solution like "we let you prime the AddressFormatter cache" that is perferable.

Reply:
Totally valid concerns you're bringing up! 

I guess the name of the function itself is really misleading as it doesn't (never did) load anything, it only _looks for_ the country inside the cache, so renaming it like `getCountryFromCache` might help clear that up a little.

## The situation: Async side-effects

A promise is a side-effect by definition in React, which leads to an in-between state while waiting, even though the data is cached for most of the app lifetime.

Here's a super simplified flow of what's happening:

1. Component renders a skeleton
2. An effect uses the async `format` function and saves the result into a state.
3. Finally replace the skeleton with the formatted address state.
4. User event triggers a change in the address, go back to step 1.

## The goal: Maximize sync flow

To avoid this forced flash of skeleton, I was in the process of managing our own cache on top of the existing one (which led me to expose a synchronous API for this package #2178), but I figured that having read access to the `AddressFormatter` cache would be best. With the cache, we can then default to a sync behavior whenever available and only fallback to async if necessary.

1. Use the sync `formatAddress` right away if the country is available in the cache.
2. If not,
   1. render skeleton
   2. let the effect load the country so that every sub-sequent renders do not need to go through another async effect.
   3. Replace skeleton with formatted address.

> people will become confused why sometimes their content load and sometimes it doesn't if they use loadCountryFromCache.

I believe the `...FromCache` part is a clear enough distinction. It's meant for an optimization use-case, if a dev only cares about getting a country, `getCountry` would be the obvious choice.

---

Priming wouldn't help since we would still _suffer_ from the async behavior we're trying to solve. And I don't want to expose write access to the cache because there's no reason to prime the cache ourselves, nor would it be safer to do so.

Reply:
After a quick meeting, here's what I'll be doing about this particular change:
- I'll revert exposing `loadCountryFromCache` to keep the API internal
- I'll explore a way to make a React binding on our side to optimize rendering synchronously whenever possible.